HIVE-26265: Option to filter out Txn events during replication.

cmunkey opened the pull request:
Purpose: Currently, all Txn events OpenTxn, CommitTxn, and RollbackTxn are included in the REPL DUMP, even when the transcation does not involve the database being dumped
(replicated). These events are unnecessary and result is excessive space required for the dump, as well as increasing work that results from these events being replayed
during REPL LOAD.

Solution proposed: To reduce this unnecessary space and work, added the hive.repl.filter.transactions configuration property. When set to "true", extra Txn events
will be filtered out as follows: CommitTxn and RollbackTxn are included in the REPL DUMP only if the transaction referenced had a corresponding ALLOCATE_WRITE_ID event
that was dumped. OpenTxn is never dumped, and the OpenTxn event will be implcitly Opened when REPL LOAD processes the ALLOC_WRITE_ID event, since the ALLOC_WRITE_ID
contains the open transaction ids. The default setting is "false".

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

pvary commented on ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/events/AbortTxnHandler.java:
```diff
@@ -39,6 +48,19 @@ public void handle(Context withinContext) throws Exception {
+       String contextDbName = StringUtils.normalizeIdentifier(withinContext.replScope.getDbName());
+       GetTxnWriteIdsRequest request = new GetTxnWriteIdsRequest(eventMessage.getTxnId());
+       request.setDbName(contextDbName);
+       GetTxnWriteIdsResponse response = withinContext.db.getMSC().getTxnWriteIds(request);
```
Could we do this without calling the metastore?
All of requested writeIds should be available during the compilation phase

cmunkey replied:
To move this during compilation, there are 2 difficulties:
1. Would need to add a new field/status to AbortTxnEvent, this field would indicate whether or not this aborted txn involved writeids. REPL DUMP could filter based on this setting.
2. The AbortTxnEvent is logged via TxnHandler.abort_txn(), so AbortTxnRequest would need to be modified to pass the writeid. AbortTxnRequest is a Thrift object. OR, abort_txn() could do the same HMS lookup that is currently done in AbortTxnHandler().

Since this metastore call is done during REPL DUMP, would it be ok to live with this inefficiency and fix later with a more optimal implementation?

pvary replied:
Running this from the client side the changes might not be atomic, which could cause issues later.
I think all of these calculations should be done on the HMS side when storing the event

cmunkey replied:
Ok, I changed how this is done.
TxnHandler.abortTxn() will check if the txn had allocated writeids, and attach the databases to the AbortTxnEvent. During REPL DUMP, when processing the AbortTxnEvent, add the event to the dump only if the AbortTxnEvent affected the database being dumped.

pvary replied:
Can we remove the HMS API change now?

cmunkey commented:
Changed the following since the last upload:

- Eliminated HMS client changes. The HMS client changes were used to query HMS to get the databases updated for a AbortTxnEvent, in the case where the event was used to notify the non transactional listeners. Currently only the DbTxnListener which is a transactional listener needs the databases updated.

- Added unit test to make sure that old versions of the AbortTxnEvent can still be serialiazed for the case where REPL DUMP may be processing events logged by an older version.

- Modified TestReplicationFilters to do extra transactions on the replica database to test out the source to target
  transaction mapping. Currently this test is causing the following problems that I am still debugging:
  
  testTxnEventsOptimized() - to assert that transaction mapping was being done correctly, I tried to check the REPL_TXN_MAP on the replica. But  this proved difficult because of the following:
  
    o REPL_TXN_MAP is cleared during commit/abort, so it could not be checked at the end of the test. Instead, I tried to capture the table in the openTxn test listener. But it seemed like the extra SELECT that the test is doing is causing some kind of deadlock in the code. Some of the sampling did work, and I could see in the debugger the mapping, after a few times sampling, the test gets a "org.apache.hadoop.hive.ql.lockmgr.LockException: Error communicating with the metastore" exception.
    o For the test to pass, I commented out the sampling of the REPL_TXN_MAP table.
    
   testTxnEventsUnoptimized() - this these was modified to burn some transactions on the replica also. But when these extra transactions are burned, on the replica, there is now extra openTxn() calls that are unexpected. This test has the transaction filtering as OFF, so it should be default behavior, so I am not sure what is going on. For the test to pass, I commented out the code to burn the transactions.

pvary commented on common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:
```diff
@@ -5613,7 +5613,12 @@ public static enum ConfVars {
-                    "tez-site.xml, etc in comma separated list.");
+                    "tez-site.xml, etc in comma separated list."),
+
+    REPL_FILTER_TRANSACTIONS("hive.repl.filter.transactions", false,
```
Maybe add something more about not dumping open transaction events?

```
"Enable transaction event filtering to save dump space.\n" +
                    "When true, open transaction evens are not dumped and transactions are implicitly opened during REPL DUMP.\n" +
                    "The default setting is false"
```

pvary commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/ReplicationTestUtils.java:
```diff
@@ -534,7 +540,7 @@ public static void assertFalseExternalFileList(WarehouseInstance warehouseInstan
 
   public static void assertExternalFileList(List<String> expected, String dumplocation,
-                                            WarehouseInstance warehouseInstance) throws IOException {
+                                            WarehouseInstance warehouseInstance) throws IOException, Exception {
```
why is this change?

cmunkey replied:
I think this is not needed.

pvary commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationFilterTransactions.java:
```diff
@@ -0,0 +1,490 @@
+      // mapping was done.
+      // But the select below seems cause the following lock error
+      //     org.apache.hadoop.hive.ql.lockmgr.LockException: Error communicating with the metastore
+//      try {
```
So we want to fix this

cmunkey replied:
If fixed this by doing a dirty read.

pvary commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:
```diff
@@ -1664,6 +1667,40 @@ public long getLatestTxnIdInConflict(long txnid) throws MetaException {
+  /**
+   * Returns the databases updated by txnId.
+   * Queries TXN_TO_WRITE_ID using txnId.
+   *
```
nit: extra line

cmunkey replied:
Done.

cmunkey commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationFilterTransactions.java:
```diff
@@ -0,0 +1,490 @@
+    Assert.assertEquals(expected.getCountCommitTxn() - filtered.getCountCommitTxn(), commitTxns.size());
+    Assert.assertEquals(expected.getCountAbortTxn() - filtered.getCountAbortTxn(), abortTxns.size());
+
+//    Map<Long, Long> replicaTxnMapping = ReplicaEventListenerTestImpl.getTxnMapping();
```
I've reenabled this.
To get TxnId skew, I do updates on the primary before dump, so all primary txnIds will be greater than
the replica txnIds.

cmunkey commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/AbortTxnEvent.java:
```diff
@@ -66,4 +75,12 @@ public Long getTxnId() {
   }
+
+  /**
+   *
```
Done.

cmunkey commented:
I was able to fix the tests.
The replication unit tests do not really allow updates to occur on the replica, this because the two DbTxnHandlers (primary and replica) actually use the same Derby database.
For the purposes of these tests, I can test the same (txnId skew) by doing extra updates on the
primary to force the primary to have larger txnIds than the replica.

pvary commented on common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:
```diff
@@ -5615,7 +5615,12 @@ public static enum ConfVars {
+
+    REPL_FILTER_TRANSACTIONS("hive.repl.filter.transactions", false,
+            "Enable transaction event filtering to save dump space.\n" +
+                    "When true, transactions are implicitly opened during REPL DUMP.\n" +
```
Not sure about the `RESTORE` - we need to use the same names as we use other filters.

```
"Enable transaction event filtering to save dump space.\n" +
        "When true, opening transactions do not generate events, and transactions on the target " +
       "server are implicitly opened during REPL RESTORE.\n" +
       "The default setting is false");
```

We want this config to be around the other `REPL` related configs, not at the end of the list.

cmunkey replied:
I will move.

pvary commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/ReplicationTestUtils.java:
```diff
@@ -17,28 +17,34 @@
  * ReplicationTestUtils - static helper functions for replication test
  */
 public class ReplicationTestUtils {
+  static final private Logger LOG = LoggerFactory.getLogger(ReplicationTestUtils.class);
```
We usually use `private static final`

cmunkey replied:
I will fixed.

pvary reviewed (approved):
Small nits, and when they are fixed we can merge this change
